Sources/MongoSwift/BSON/BSONValue.swift

```diff
@@ -103,7 +103,7 @@ extension Array: BSONValue {
     public func encode(to storage: DocumentStorage, forKey key: String) throws {
         // An array is just a document with keys '0', '1', etc. corresponding to indexes
         var arr = Document()
-        for (i, v) in self.enumerated() { arr[String(i)] = v as? BSONValue }
+        for (i, v) in self.enumerated() { try arr.setValue(forKey: String(i), to: v as? BSONValue) }
```

Comment:
while you're here: note that this method will silently fail and encode a `nil` when the Array contains a non-`BsonValue?`, as the `as? BSONValue` cast just returns nil.
so if I had an array like `[1, "hi", SomeNonBSONValueType(), 3]`, it ends up getting encoded as `[1, "hi", nil, 3]`. 

Since this method throws now anyway (it didn't use to), how about throwing an error here when v is neither `nil` nor castable to a `BSONValue`? 

the downside of that is that code would now crash if you tried to put such an array in a document via th subscript setter. I'm not sure whether this silent unexpected behavior is preferable to a crash... maybe @jmikola will have thoughts as well.

Reply:
How is this case possible if arrays in Swift are homogenous? The only way this can happen from what I can tell if we are playing with an array of type `[Any]`. However, with [SWIFT-215](https://jira.mongodb.org/browse/SWIFT-215), this should not happen right?

In either case, since we may not know when SWIFT-215 will be done, it may still be better to do something explicit here instead of silently fail. My own opinion almost always about this sort of stuff is being explicit and annoying the user away from doing dumb or bad things is better than being quiet.

Reply:
yup, that's correct, it could only be an `[Any]` or `[Non-BSONValue]`, and indeed, SWIFT-215 will solve that once we drop 4.0.
I'm good with throwing an error here in those cases. we should probably also add an explicit note to the docstring for the Array extension noting that it's currently on the user to ensure their arrays only contain `BSONValue`s.

Reply:
Sounds good to me as well. Since this seems to be the only thing, I'll wait to see if @jmikola has any thoughts on this as well.

Reply:
>  how about throwing an error here when v is neither nil nor castable to a BSONValue?

Would this be done by checking if `v` is first nil and, if not, catching if the result of `v as? BSONValue` yields nil?

I don't quite follow how this would result in a crash. Is the subscript setter using `encode()` and raising a precondition failure if an exception is thrown?

Reply:
@jmikola I don't think the issue is related to the new code crashing. The issue is that it will silently give `nil` for non-`BSONValue`'s. What we are proposing however is to throw something if `v` is not a `BSONValue`.

If you're asking about how the old code would have crashed, then yes, the current subscript setter will raise a `preconditionFailure()` whenever `setValue()` throws. `setValue()` eventually calls `encode()` (if it is a new key, which occurs every time in this case since we are starting with a new `Document`).

> Would this be done by checking if v is first nil and, if not, catching if the result of v as? BSONValue yields nil?

And yes, I think that is how it would be done.

Let me know if I misinterpreted you.

Reply:
yes, it would lead to a crash because the subscript setter calls through to this function.

Reply:
If the current code would raise a precondition failure when encoding failed, I'm all for keeping that here. My instincts tell me it's worse to silently encode a nil and hide errors from the user.

I don't have the bandwidth to think about it now, but we'd probably do well to revisit all of this and audit the driver to places where users might experience precondition failures due to runtime errors. Then we can properly document all of those cases and chalk them up to necessary evils due to language limitations (i.e. not being able to throw from certain contexts).

Reply:
yeah, it was still possible if something else went wrong in encoding; this is just adding another possible error that could result in a crash. agreed on the audit

Reply:
let's definitely not revert, we want to throw from here and have the precondition failure happen at the top-level subscript `set` call rather than in this recursive call (since subscript `set` calls this method...) we were silent encoding nil with the old way too. let's also add an error in the non-nil case. 
as an interesting side note, technically we could support anything Codable being in an array, not just anything that's a BSONValue. because Codable things can be turned into Documents via BSONEncoder, and are therefore kinda BSONValues themselves... 🤯

Reply:
@kmahar I've added this detail about `Codable` to [SWIFT-215](https://jira.mongodb.org/browse/SWIFT-215).

Also, we now throw in the case that we get a non-`BSONValue`.